activerecord/lib/active_record/relation/batches.rb

```diff
@@ -287,15 +287,38 @@ def apply_limits(relation, start, finish, order)
       end
 
       def apply_start_limit(relation, start, order)
-        relation.where(predicate_builder[primary_key, start, order == :desc ? :lteq : :gteq])
+        batch_condition(relation, primary_key, start, order == :desc ? :lteq : :gteq)
       end
 
       def apply_finish_limit(relation, finish, order)
-        relation.where(predicate_builder[primary_key, finish, order == :desc ? :gteq : :lteq])
+        batch_condition(relation, primary_key, finish, order == :desc ? :gteq : :lteq)
+      end
+
+      def batch_condition(relation, columns, values, operator)
+        columns = Array(columns)
+        values = Array(values)
+        last_index = columns.size - 1
+        where_clause = nil
+
+        last_index.downto(0) do |column_index|
+          column = columns[column_index]
+          value = values.fetch(column_index)
+
+          where_clause =
+            if column_index == last_index
+              predicate_builder[column, value, operator]
+            else
+              predicate_builder[column, value, operator == :lteq ? :lt : :gt].or(
+                predicate_builder[column, value, :eq].and(where_clause)
+              )
+            end
+        end
+
+        relation.where(where_clause)
       end
 
       def batch_order(order)
-        table[primary_key].public_send(order)
+        Array(primary_key).map { |column| table[column].public_send(order) }
```

Comment:
Here we generally have an option to either branch on `composite_primary_key?` to avoid unnecessary allocation or do the wrap for the purpose of unifying the code and so far we've been opting into doing the branching. However I believe at this point we should consider adding an internal concept called something like `composite_primary_key` which will be memoized on the model class and will always be an array, including single column PKs like `[:id]` so we can unify the code without allocating array more than once.

So I think we can keep the wrap with a plan to refactor all wraps (and branches) into using this memoized always-an-array version of the primary key

Reply:
In this case, this method is called only once at the start of the batching. But yeah, `batch_condition`, which is called on each iteration, allocates the same array over and over. Not a big deal for me, because batching is already a heavy operation due to lots of SQL, at least.

Maybe better to introduce `composite_primary_key?`/`composite_primary_key` in a separate PR, if some existing code would already benefit from it? And then use it there. Or I can do it in this PR, if not.